### PR TITLE
Reduced image size for demo and builds

### PR DESCRIPTION
**Pull request by wsobel**
* Reduced the docker image size for demo and agent builds.

### CODE REVIEW

**Review by simonyg — APPROVED** [review]

File: demo/Dockerfile [hunk 1]
```diff
@@ -30,6 +30,8 @@
 # 22.04 is the current long term support release, maintained until 2025-04.
 FROM ubuntu:22.04 AS os
 
+RUN apt-get update
```
Comment (ottobolyos) [inline demo/Dockerfile:33]: Why did you move this into a separate `RUN`? 🤔

File: demo/Dockerfile [hunk 2]
```diff
@@ -70,11 +71,10 @@ RUN  apt-get update \
...
 
 # bring in the repo contents, minus .dockerignore files
-COPY . .
+COPY . cppagent
 
 ARG WITH_TESTS=false
 ARG WITH_TESTS_ARG=argument
-ARG SHARED=False
 
 # Build and optionally test
```
Comment (ottobolyos) [inline demo/Dockerfile:79]: IMHO we could make the two dockerfiles identical apart from OS differences e.g. deps installation).

```suggestion
# Build and optionally test. Unpack the dist to reduce overhead in the following release step.
```

File: demo/Dockerfile [hunk 3]
```diff
@@ -83,6 +83,7 @@ RUN  if [ -z "$WITH_TESTS" ] || [ "$WITH_TESTS" = "false" ]; then \
        WITH_TESTS_ARG=""; \
      fi \
   && conan profile detect \
+  && cd cppagent \
   && conan create . \
```
Comment (ottobolyos) [inline demo/Dockerfile:87]: I believe we could eliminate changing directory by replacing `.` with an absolute path.

```suggestion
  && cd cppagent \
  && conan create /root/agent/cppagent \
```

File: demo/Dockerfile [hunk 4]
```diff
@@ -93,7 +94,10 @@ RUN  if [ -z "$WITH_TESTS" ] || [ "$WITH_TESTS" = "false" ]; then \
        -o cpack_name=dist \
        -o cpack_generator=TGZ \
        -pr "$CONAN_PROFILE" \
-       ${WITH_TESTS_ARG}
+       ${WITH_TESTS_ARG} \
```
Comment (ottobolyos) [inline demo/Dockerfile:97]: Nitpick.

```suggestion
       "$WITH_TESTS_ARG" \
```

File: demo/Dockerfile [hunk 5]
```diff
@@ -93,7 +94,10 @@ RUN  if [ -z "$WITH_TESTS" ] || [ "$WITH_TESTS" = "false" ]; then \
...
+  && cd .. && tar xf dist.tar.gz \
+  && rm dist.tar.gz
```
Comment (ottobolyos) [inline demo/Dockerfile:99]: No need to change directory.

```suggestion
  && tar xf /root/agent/dist.tar.gz -C /root/agent \
  && rm /root/agent/dist.tar.gz
```

File: demo/Dockerfile [hunk 6]
```diff
@@ -93,7 +94,10 @@ RUN  if [ -z "$WITH_TESTS" ] || [ "$WITH_TESTS" = "false" ]; then \
...
+
```
Comment (ottobolyos) [inline demo/Dockerfile:100]: Nitpick: Any reason behind adding the blank line? 🤔

```suggestion
```

File: demo/Dockerfile.alpine [hunk 7]
```diff
@@ -17,7 +16,7 @@
 #   To run tests, use `--build-arg WITH_TESTS=true`
 #
 #   # Note: In this case, I would suggest to map port `5000` to `5000`. The user can always change the port according to their needs.
-#   docker run -it --rm --init --name agent -p5000:5000 \
+#   docker run -it --rm --init --name agent -p5000:5000 -v ... \
```
Comment (ottobolyos) [inline demo/Dockerfile.alpine:19]: I’d suggest to add an example volume binding both here and in `Dockerfile`, unless you’d like to move whole note to the docs. I suggest to do this in all four dockerfiles.

File: demo/Dockerfile.alpine [hunk 8]
```diff
@@ -17,7 +16,7 @@
...
 #     mtconnect/agent:2.0.0.12_RC18
```
Comment (ottobolyos) [inline demo/Dockerfile.alpine:20]: I suggext to remove the version tag unless you want to update in on every release, unless you want to move the whole notes in the docs (then it should be done there). I suggest to do this in all four dockerfiles.

```suggestion
#     mtconnect/agent
```

File: demo/Dockerfile.alpine [hunk 9]
```diff
@@ -29,6 +28,8 @@
 # base image - alpine 3.18
 FROM alpine:3.18 AS os
 
+RUN apk update
```
Comment (ottobolyos) [inline demo/Dockerfile.alpine:31]: Why did you move this into a separate `RUN`? 🤔

Anyway, we could also use `apk --update add` below and remove this command altogether.

File: demo/Dockerfile.alpine [hunk 10]
```diff
@@ -63,24 +63,25 @@ RUN  apk update \
...
-# Build and optionally test
+# Build and optionally test. Unpack the dist to reduce overhead in the following release step.
 RUN  if [ -z "$WITH_TESTS" ] || [ "$WITH_TESTS" = "false" ]; then \
        WITH_TESTS_ARG="--test-folder="; \
      else \
        WITH_TESTS_ARG=""; \
      fi \
   && conan profile detect \
+  && cd cppagent \
   && conan create . \
```
Comment (ottobolyos) [inline demo/Dockerfile.alpine:85]: I believe we could eliminate changing directory by replacing `.` with an absolute path.

```suggestion
  && cd cppagent \
  && conan create /root/agent/cppagent \
```

File: demo/Dockerfile.alpine [hunk 11]
```diff
@@ -91,95 +92,83 @@ RUN  if [ -z "$WITH_TESTS" ] || [ "$WITH_TESTS" = "false" ]; then \
        -o cpack_name=dist \
        -o cpack_generator=TGZ \
        -pr "$CONAN_PROFILE" \
-       ${WITH_TESTS_ARG}
+       ${WITH_TESTS_ARG} \
```
Comment (ottobolyos) [inline demo/Dockerfile.alpine:95]: Nitpick.

```suggestion
       "$WITH_TESTS_ARG" \
```

File: demo/Dockerfile.alpine [hunk 12]
```diff
@@ -91,95 +92,83 @@ RUN  if [ -z "$WITH_TESTS" ] || [ "$WITH_TESTS" = "false" ]; then \
...
+  && cd .. && tar xf dist.tar.gz \
+  && rm dist.tar.gz
```
Comment (ottobolyos) [inline demo/Dockerfile.alpine:97]: No need to change directory.

```suggestion
  && tar xf /root/agent/dist.tar.gz -C /root/agent \
  && rm /root/agent/dist.tar.gz
```

File: docker/alpine/Dockerfile [hunk 13]
```diff
@@ -29,6 +28,8 @@
 # base image - alpine 3.18
 FROM alpine:3.18 AS os
 
+RUN apk update
```
Comment (ottobolyos) [inline docker/alpine/Dockerfile:31]: Why did you move this into a separate `RUN`? 🤔

File: docker/alpine/Dockerfile [hunk 14]
```diff
@@ -69,18 +69,19 @@ RUN  apk update \
...
-# Build and optionally test
+# Build and optionally test. Unpack the dist to reduce overhead in the following release step.
 RUN  if [ -z "$WITH_TESTS" ] || [ "$WITH_TESTS" = "false" ]; then \
        WITH_TESTS_ARG="--test-folder="; \
      else \
        WITH_TESTS_ARG=""; \
      fi \
   && conan profile detect \
+  && cd cppagent \
   && conan create . \
```
Comment (ottobolyos) [inline docker/alpine/Dockerfile:85]: I believe we could eliminate changing directory by replacing `.` with an absolute path.

```suggestion
  && cd cppagent \
  && conan create /root/agent/cppagent \
```

File: docker/alpine/Dockerfile [hunk 15]
```diff
@@ -91,74 +92,68 @@ RUN  if [ -z "$WITH_TESTS" ] || [ "$WITH_TESTS" = "false" ]; then \
        -o cpack_name=dist \
        -o cpack_generator=TGZ \
        -pr "$CONAN_PROFILE" \
-       ${WITH_TESTS_ARG}
+       ${WITH_TESTS_ARG} \
```
Comment (ottobolyos) [inline docker/alpine/Dockerfile:95]: Nitpick.

```suggestion
       "$WITH_TESTS_ARG" \
```

File: docker/alpine/Dockerfile [hunk 16]
```diff
@@ -91,74 +92,68 @@ RUN  if [ -z "$WITH_TESTS" ] || [ "$WITH_TESTS" = "false" ]; then \
...
+  && cd .. && tar xf dist.tar.gz \
+  && rm dist.tar.gz
```
Comment (ottobolyos) [inline docker/alpine/Dockerfile:97]: No need to change directory.

```suggestion
  && tar xf /root/agent/dist.tar.gz -C /root/agent \
  && rm /root/agent/dist.tar.gz
```

File: docker/ubuntu/Dockerfile [hunk 17]
```diff
@@ -30,6 +30,8 @@
 # 22.04 is the current long term support release, maintained until 2025-04.
 FROM ubuntu:22.04 AS os
 
+RUN apt-get update
```
Comment (ottobolyos) [inline docker/ubuntu/Dockerfile:33]: Why did you move this into a separate `RUN`? 🤔

File: docker/ubuntu/Dockerfile [hunk 18]
```diff
@@ -82,6 +83,7 @@ RUN  if [ -z "$WITH_TESTS" ] || [ "$WITH_TESTS" = "false" ]; then \
        WITH_TESTS_ARG=""; \
      fi \
   && conan profile detect \
+  && cd cppagent \
   && conan create . \
```
Comment (ottobolyos) [inline docker/ubuntu/Dockerfile:87]: I believe we could eliminate changing directory by replacing `.` with an absolute path.

```suggestion
  && cd cppagent \
  && conan create /root/agent/cppagent \
```

File: docker/ubuntu/Dockerfile [hunk 19]
```diff
@@ -92,7 +94,10 @@ RUN  if [ -z "$WITH_TESTS" ] || [ "$WITH_TESTS" = "false" ]; then \
        -o cpack_name=dist \
        -o cpack_generator=TGZ \
        -pr "$CONAN_PROFILE" \
-       ${WITH_TESTS_ARG}
+       ${WITH_TESTS_ARG} \
```
Comment (ottobolyos) [inline docker/ubuntu/Dockerfile:97]: Nitpick.

```suggestion
       "$WITH_TESTS_ARG" \
```

File: docker/ubuntu/Dockerfile [hunk 20]
```diff
@@ -92,7 +94,10 @@ RUN  if [ -z "$WITH_TESTS" ] || [ "$WITH_TESTS" = "false" ]; then \
...
+  && cd .. && tar xf dist.tar.gz \
+  && rm dist.tar.gz
```
Comment (ottobolyos) [inline docker/ubuntu/Dockerfile:99]: No need to change directory.

```suggestion
  && tar xf /root/agent/dist.tar.gz -C /root/agent \
  && rm /root/agent/dist.tar.gz
```

File: docker/ubuntu/Dockerfile [hunk 21]
```diff
@@ -92,7 +94,10 @@ RUN  if [ -z "$WITH_TESTS" ] || [ "$WITH_TESTS" = "false" ]; then \
...
+
```
Comment (ottobolyos) [inline docker/ubuntu/Dockerfile:100]: Nitpick: Any reason behind adding the blank line? 🤔

```suggestion
```

**Review by ottobolyos — CHANGES_REQUESTED** [review]
@wsobel, I have some suggestions and nitpicks to update the changed file. Overall, it looks great! :+1:

File: demo/Dockerfile (hunk 3 above)
Comment (wsobel) [reply demo/Dockerfile:87]: I'll make the change

File: demo/Dockerfile (hunk 4 above)
Comment (wsobel) [reply demo/Dockerfile:97]: Makes sense

File: demo/Dockerfile (hunk 5 above)
Comment (wsobel) [reply demo/Dockerfile:99]: Same as previous. Good point.

File: demo/Dockerfile.alpine (hunk 7 above)
Comment (wsobel) [reply demo/Dockerfile.alpine:19]: Ok

File: demo/Dockerfile.alpine (hunk 8 above)
Comment (wsobel) [reply demo/Dockerfile.alpine:20]: Agree. I need to do it in the workflow. It will required a little sed script to change the ref name.

File: demo/Dockerfile.alpine (hunk 9 above)
Comment (wsobel) [reply demo/Dockerfile.alpine:31]: No good reason. Just to consolidate common code. I can mov it back since it will create another layer.

File: demo/Dockerfile (hunk 2 above)
Comment (ottobolyos) [reply demo/Dockerfile]: Indeed, that is a good idea. :+1: What language will you use? :thinking: If Bash or JavaScript/TypeScript, I could help you out if you want. :wink:

File: demo/Dockerfile (hunk 3 above)
Comment (ottobolyos) [reply demo/Dockerfile:87]: Okay. Now I see that I created wrong suggestion, as I selected only one line instead of two, sorry. :man_facepalming:

File: demo/Dockerfile [hunk 22]
```diff
@@ -70,19 +71,19 @@ RUN  apt-get update \
...
-# Build and optionally test
+# Build and optionally test. Unpack the dist to reduce overhead in the following release step.
 RUN  if [ -z "$WITH_TESTS" ] || [ "$WITH_TESTS" = "false" ]; then \
        WITH_TESTS_ARG="--test-folder="; \
      else \
        WITH_TESTS_ARG=""; \
      fi \
   && conan profile detect \
+  && cd cppagent \
   && conan create . \
```
Comment (ottobolyos) [inline demo/Dockerfile:87]:
```suggestion
  && conan create /root/agent/cppagent \
```

File: docker/ubuntu/Dockerfile (hunk 18 above)
Comment (ottobolyos) [reply docker/ubuntu/Dockerfile:87]: Again, I added wrong suggestion, here’s the corrected one:

```suggestion
  && conan create /root/agent/cppagent \
```

File: docker/ubuntu/Dockerfile (hunk 19 above)
Comment (ottobolyos) [reply docker/ubuntu/Dockerfile:97]: You have marked this as resolved, however, it seems not to be changed in the PR. Did you commit all your changes? :thinking:

**Comment by wsobel** [comment]
I'm going to change all the demo Dockerfiles to use the mtconnect/agent:latest as the base. This will remove the maintenance headache I'm currently having. I will only maintain the ubuntu and alpine base sets, all others will be derivative.

File: docker/ubuntu/Dockerfile (hunk 17 above)
Comment (wsobel) [reply docker/ubuntu/Dockerfile:33]: Changed it

File: demo/Dockerfile (hunk 2 above)
Comment (wsobel) [reply demo/Dockerfile]: I'm not sure if I will generate. Since we have the dockerhub account and the base image, I'll just use that as the foundation.

Comment (wsobel) [reply demo/Dockerfile]: I'll generate when I have more time.

File: demo/Dockerfile.alpine (hunk 8 above)
Comment (wsobel) [reply demo/Dockerfile.alpine:20]: Changed to `#     --tag mtconnect/agent:latest \`

File: demo/Dockerfile.alpine (hunk 10 above)
Comment (wsobel) [reply demo/Dockerfile.alpine:85]: we could also use relative since the workdir is already set

File: docker/ubuntu/Dockerfile (hunk 17 above)
Comment (ottobolyos) [reply docker/ubuntu/Dockerfile:33]: Where? :thinking:

File: demo/Dockerfile.alpine (hunk 8 above)
Comment (ottobolyos) [reply demo/Dockerfile.alpine:20]: I just wonder if `--tag` is actually needed. :thinking:

File: demo/Dockerfile.alpine (hunk 10 above)
Comment (ottobolyos) [reply demo/Dockerfile.alpine:85]: Indeed, we could; like this:

```suggestion
  && conan create cppagent \
```

**Comment by ottobolyos** [comment]
> I'm going to change all the demo Dockerfiles to use the mtconnect/agent:latest as the base.

Actually, do we need those demo dockerfiles at all? Can’t simply provide Docker Compose files where would create a service from `mtconnect/agent` image and bind-mount the required volumes? :thinking:

**Comment by wsobel** [comment]
> > I'm going to change all the demo Dockerfiles to use the mtconnect/agent:latest as the base.
> 
> Actually, do we need those demo dockerfiles at all? Can’t simply provide Docker Compose files where would create a service from `mtconnect/agent` image and bind-mount the required volumes? 🤔

I have two compose files right now that do what this. One without MQTT and one with MQTT. I'm going to change them to use the dockerhub version of the agent. 

I thought a straight Docker image may be useful for people who would like an example of a non-composed version. I get your point and was debating it myself. I hate maintaining extra stuff.

**Comment by wsobel** [comment]
Finished changes and made demo and replicator dependent on the mtconnect/agent dockerfile.

Removed the leading v from the version (needs to be tested on next tagging)

**Comment by wsobel** [comment]
All issues addressed. Can still make small style changes if necessary.

File: demo/Dockerfile [hunk 23]
```diff
@@ -10,193 +10,75 @@
...
 #     --platform linux/amd64,linux/arm64 \
-#     --tag mtconnect/agent:2.0.0.12_RC18 \
+#     --tag mtconnect/demo:latest \
 #     --push \
+#     -f demo/Dockerfile
 #     .
 #
 #   To run tests, use `--build-arg WITH_TESTS=true`
 #
 #   # Note: In this case, I would suggest to map port `5000` to `5000`. The user can always change the port according to their needs.
```
Comment (ottobolyos) [inline demo/Dockerfile:20]: Remove this comment please.

```suggestion
```

File: demo/Dockerfile [hunk 24]
```diff
@@ -10,193 +10,75 @@
...
-RUN  apt-get update \
-  && apt-get install -y ruby
+USER root
 
-# change to a new non-root user for better security.
-# this also adds the user to a group with the same name.
-# -m creates a home folder, ie /home/<username>
-ARG UID=1000
-ARG GID=1000
+RUN apk add ruby
```
Comment (ottobolyos) [inline demo/Dockerfile:35]: I think you have forgotten to update APK’s cache. You can either `-U` or `--update` or `--update-cache` with `apk add` subcommand.

```suggestion
RUN apk --update add ruby
```

File: demo/compose/compose.yaml [hunk 25]
```diff
@@ -25,5 +11,15 @@ services:
       dockerfile: demo/compose/adapter/Dockerfile
       args:
         - LOG_FILE=okuma.txt
-
-    
+  agent:
+    image: "mtconnect/agent:latest"
+    volumes:
+      - ./agent/agent.cfg:/mtconnect/config/agent.cfg
+      - ../agent/Devices.xml:/mtconnect/config/Devices.xml
```
Comment (ottobolyos) [inline demo/compose/compose.yaml:18]: Do we need to bind-mount `agent.cfg` and `Devices.xml` when we copy them into the image at build time? :thinking:

File: docker/alpine/Dockerfile [hunk 26]
```diff
@@ -9,6 +8,7 @@
 # (see CMakeLists.txt for current version number) -
 #
 #   docker buildx build \
+#     -f docker/ubuntu/Dockerfile \
 #     --platform linux/amd64,linux/arm64 \
 #     --tag mtconnect/agent:2.0.0.12_RC18 \
```
Comment (ottobolyos) [inline docker/alpine/Dockerfile:13]: Either remove `2.0.0.12_RC18` or replace it with `latest` or update it on each release.

```suggestion
#     --tag mtconnect/agent:latest \
```

File: docker/alpine/Dockerfile [hunk 27]
```diff
@@ -17,8 +17,9 @@
 #   To run tests, use `--build-arg WITH_TESTS=true`
 #
 #   # Note: In this case, I would suggest to map port `5000` to `5000`. The user can always change the port according to their needs.
```
Comment (ottobolyos) [inline docker/alpine/Dockerfile:19]: Remove this comment.

```suggestion
```

File: docker/alpine/Dockerfile [hunk 28]
```diff
@@ -17,8 +17,9 @@
...
-#   docker run -it --rm --init --name agent -p5000:5000 \
-#     mtconnect/agent:2.0.0.12_RC18
+#   docker run -it --rm --init --name agent -p5000:5000 -v ./agent/agent.cfg:/mtconnect/config/agent.cfg \
+#        -v ./agent/Devices.xml:/mtconnect/config/Devices.xml \
+#        mtconnect/agent:2.0.0.12_RC18
```
Comment (ottobolyos) [inline docker/alpine/Dockerfile:22]: Either remove `2.0.0.12_RC18` or replace it with `latest` or update it on each release.

```suggestion
#        mtconnect/agent:latest
```

File: docker/alpine/Dockerfile [hunk 29]
```diff
@@ -91,74 +91,69 @@ RUN  if [ -z "$WITH_TESTS" ] || [ "$WITH_TESTS" = "false" ]; then \
...
 # change to a new non-root user for better security.
 # this also adds the user to a group with the same name.
 # --create-home creates a home folder, ie /home/<username>
 ARG UID=1000
 ARG GID=1000
 
-RUN  addgroup \
+# install dependencies
+RUN  apk update \
+  && apk add \
```
Comment (ottobolyos) [inline docker/alpine/Dockerfile:114]: Consider replacing this with the following:

```suggestion
RUN apk --update add \
```

File: docker/replicator/Dockerfile [hunk 30]
```diff
@@ -10,151 +9,34 @@
...
+#     --tag mtconnect/replicator:latest \
 #     --push \
 #     .
 #
-#   To run tests, use `--build-arg WITH_TESTS=true`
-#
 #   # Note: In this case, I would suggest to map port `5000` to `5000`. The user can always change the port according to their needs.
-#   docker run -it --rm --init --name agent -p5000:5000 \
-#     mtconnect/agent:2.0.0.12_RC18
+#   docker run -it --rm --init --name replicator -p5000:5000 -e SOURCE_URL=http://192.168.5.100:5000/ \
```
Comment (ottobolyos) [inline docker/replicator/Dockerfile:17]: Consider updating `SOURCE_URL`. What should be its value? :thinking:

File: docker/ubuntu/Dockerfile [hunk 31]
```diff
@@ -9,16 +9,18 @@
...
+#     -f docker/ubuntu/Dockerfile \
 #     --platform linux/amd64,linux/arm64 \
-#     --tag mtconnect/agent:2.0.0.12_RC18 \
+#     --tag mtconnect/agent:latest \
 #     --push \
 #     .
 #
 #   To run tests, use `--build-arg WITH_TESTS=true`
 #
 #   # Note: In this case, I would suggest to map port `5000` to `5000`. The user can always change the port according to their needs.
```
Comment (ottobolyos) [inline docker/ubuntu/Dockerfile:20]: Remove this comment.

```suggestion
```

File: src/mtconnect/sink/rest_sink/file_cache.cpp [hunk 32]
```diff
@@ -311,43 +311,47 @@ namespace mtconnect::sink::rest_sink {
...
+            m_fileCache.insert_or_assign(name, file);
+          }
+          else
+          {
+            file = findFileInDirectories(name);
+          }
         }
       }
-
+
```
Comment (ottobolyos) [inline src/mtconnect/sink/rest_sink/file_cache.cpp:354]: Useless trailing whitespace. Consider configuring your IDE to remove traling whitespace on save (e.g. using [`.editorconfig`](https://editorconfig.org/)).

**Review by ottobolyos — COMMENTED** [review]
@wsobel, all I have are some nitpicks.

File: demo/Dockerfile (hunk 24 above)
Comment (wsobel) [reply demo/Dockerfile:35]: Update was done before on the other image. I can add it in.

File: demo/compose/compose.yaml (hunk 25 above)
Comment (wsobel) [reply demo/compose/compose.yaml:18]: These are different files, not the ones that are in the image. This is not using the demo image, it's the base image.

File: docker/replicator/Dockerfile (hunk 30 above)
Comment (wsobel) [reply docker/replicator/Dockerfile:17]: It depends on what the source URL is. Don't get the point?

File: docker/ubuntu/Dockerfile (hunk 31 above)
Comment (wsobel) [reply docker/ubuntu/Dockerfile:20]: done

File: docker/alpine/Dockerfile (hunk 28 above)
Comment (wsobel) [reply docker/alpine/Dockerfile:22]: removed version, latest is default

File: docker/alpine/Dockerfile (hunk 29 above)
Comment (wsobel) [reply docker/alpine/Dockerfile:114]: changed

File: src/mtconnect/sink/rest_sink/file_cache.cpp (hunk 32 above)
Comment (wsobel) [reply src/mtconnect/sink/rest_sink/file_cache.cpp:354]: not worth it.

**Comment by wsobel** [comment]
Createed a new branch for a new PR.
